source/common/network/utility.cc

```diff
@@ -240,16 +240,32 @@ Address::InstanceConstSharedPtr Utility::getLocalAddress(const Address::IpVersio
 
 bool Utility::isLocalConnection(const Network::ConnectionSocket& socket) {
   const auto& remote_address = socket.remoteAddress();
-  // Before calling getifaddrs, verify the obvious checks.
-  // Note that there are corner cases, where remote and local address will be the same
-  // while the client is not actually local. Example could be an iptables intercepted
-  // connection. However, this is a rare exception and such assumption results in big
-  // performance optimization.
+  // Before calling getifaddrs, verify the obvious checks. These are local:
+  // - Pipes
+  // - Sockets to a loopback address
+  // - Sockets where the local and remote address (ignoring port) are the same
   if (remote_address->type() == Envoy::Network::Address::Type::Pipe ||
-      remote_address == socket.localAddress() || isLoopbackAddress(*remote_address)) {
+      isLoopbackAddress(*remote_address)) {
     return true;
   }
+  const auto local_ip = socket.localAddress()->ip();
```

Comment:
I think you can simplify this code by just comparing remote_address->ip()->addressAsString() == local_address->ip()->addressAsString(). If we deeply care about speed, I might go ahead and add an equality comparison on the Ip interface, and then deal with it inside there so we don't duplicate this code elsewhere?

Reply:
addressAsString() is fine with me.

Equality comparison on Ip can be done, but I have some hesitation.  I think that means adding `operator==() PURE` to [Envoy::Network::Address::Ip](https://github.com/envoyproxy/envoy/blob/master/include/envoy/network/address.h#L53), and for my purposes, making it compare IPs but *not ports*, even though port is a member of the Ipv4/Ipv6 implementation.  But, Envoy::Network::Address::Instance would have `operator==() PURE` that for IPs, compares it but [*with ports*](https://github.com/envoyproxy/envoy/blob/master/source/common/network/address_impl.cc#L217).

If I was a programmer that just showed up, I would be surprised if `*address1 == *address2` did compare ports, but `*address1->ip() == *address2->ip()` did not compare ports, especially since the port is a property of `*address1->ip()` (so `*address1->ip()->port() != *address2->ip()->port()`).  Basically I think since port is a property of the IP address, it's gotta be reflected in all `operator==()` up and down the class hierarchy the same.

My complaint though is a style consideration and I'm new here so I'm happy to do it if you're OK with it.

Other options:
1. Make the equality comparison not named `operator==()`, make it some other method (`addressEquals()` on the Ip class is a better name than `equalsExceptPort()` on Address).
2. Remove `port` from the Address.  Maybe ports are on some new thing, maybe Sockaddrs, but not Ip.  Feels like this is a pretty big change.

I'll update to use `addressAsString()` so we can take a look.  I'm OK with any of these choices though.

Reply:
Hmm, I just looked and `addressAsString()` includes port, so I don't think that is going to work either?

Maybe just factor that code out into a utility function that we can test separately? You could make a function that takes 2 `Address::Ip` and returns whether they are equal without ports? WDYT?

Reply:
Turns out that `addressAsString()` does not include port, it's only the IP address.  `address->asString()` does include the IP and port.  Pushed code that works using `addressAsString()` (and double-checked locally that if I mangle an address but not a port, the string comparison fails as expected).

Gory details: `addressAsString()` returns the private member `friendly_address_` which is initialized with only the IPv4/IPv6 as a string (excluding port).  `friendly_name_` includes the IPv4/IPv6 **and** the port.

https://github.com/envoyproxy/envoy/blob/cb15cc3d2010aff77d6e022ddf6d723fa8becbc0/source/common/network/address_impl.cc#L178-L189

So, the code that I just pushed uses `addressAsString()` and works correctly.  I am also not terrified of the performance cost of string comparison especially since we are already always constructing the strings at `Ipv4`/`Ipv6` construction time.